node/forest_libp2p/src/service.rs

```diff
@@ -1,42 +1,68 @@
 // Copyright 2020 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-use super::behaviour::{MyBehaviour, MyBehaviourEvent};
+use super::behaviour::{ForestBehaviour, ForestBehaviourEvent};
 use super::config::Libp2pConfig;
-use futures::{Async, Stream};
+use async_std::sync::{channel, Receiver, Sender};
+use futures::select;
+use futures_util::stream::StreamExt;
 use libp2p::{
     core,
     core::muxing::StreamMuxerBox,
     core::nodes::Substream,
     core::transport::boxed::Boxed,
-    gossipsub::TopicHash,
+    gossipsub::{Topic, TopicHash},
     identity::{ed25519, Keypair},
     mplex, secio, yamux, PeerId, Swarm, Transport,
 };
 use slog::{debug, error, info, trace, Logger};
 use std::io::{Error, ErrorKind};
 use std::time::Duration;
 use utils::{get_home_dir, read_file_to_vec, write_to_file};
+
 type Libp2pStream = Boxed<(PeerId, StreamMuxerBox), Error>;
-type Libp2pBehaviour = MyBehaviour<Substream<StreamMuxerBox>>;
+type Libp2pBehaviour = ForestBehaviour<Substream<StreamMuxerBox>>;
+
+/// Events emitted by this Service
+#[derive(Clone, Debug)]
+pub enum NetworkEvent {
+    PubsubMessage {
+        source: PeerId,
+        topics: Vec<TopicHash>,
+        message: Vec<u8>,
+    },
+}
 
+/// Events into this Service
+#[derive(Clone, Debug)]
+pub enum NetworkMessage {
+    PubsubMessage { topic: Topic, message: Vec<u8> },
+}
 /// The Libp2pService listens to events from the Libp2p swarm.
 pub struct Libp2pService {
-    pub swarm: Swarm<Libp2pStream, Libp2pBehaviour>,
+    swarm: Swarm<Libp2pStream, Libp2pBehaviour>,
+
+    pubsub_receiver_in: Receiver<NetworkMessage>,
+    pubsub_sender_in: Sender<NetworkMessage>,
+
+    pubsub_receiver_out: Receiver<NetworkEvent>,
+    pubsub_sender_out: Sender<NetworkEvent>,
+
+    log: Logger,
 }
 
 impl Libp2pService {
     /// Constructs a Libp2pService
-    pub fn new(log: &Logger, config: &Libp2pConfig) -> Self {
-        let net_keypair = get_keypair(log);
+    pub fn new(log: Logger, config: &Libp2pConfig) -> Self {
+        let net_keypair = get_keypair(&log);
         let peer_id = PeerId::from(net_keypair.public());
 
         info!(log, "Local peer id: {:?}", peer_id);
 
         let transport = build_transport(net_keypair.clone());
 
         let mut swarm = {
-            let be = MyBehaviour::new(log.clone(), &net_keypair);
+            let be = ForestBehaviour::new(log.clone(), &net_keypair);
             Swarm::new(transport, be, peer_id)
         };
 
```

Comment:
Although not changed in this PR, should we avoid unwrapping on `Swarm::liston_on` on LN 86 (below). 

I don't see why Github does not allow to comment on code unchanged in pr reviews, saw it being requested [here as a feature](https://github.com/isaacs/github/issues/284) interestingly.

Reply:
I feel like yall commented on it last time. And I don’t see how you can handle that kinda error. If you cant even listen_on, we shouldn’t handle that